include/boost/numeric/ublas/tensor/detail/basic_type_traits.hpp

```diff
@@ -60,6 +64,14 @@ struct is_static_rank : std::false_type {};
 template <class E> 
 inline static constexpr bool const is_static_rank_v = is_static_rank<E>::value;
 
+
+template<typename>
+struct always_false : std::false_type{};
+
+
+template <typename T> 
+inline static constexpr bool const always_false_v = always_false<T>::value;
+
```

Comment:
The problem with `std::false_type` is it does not delay the `static_assert`. If I use `std::false_type` `static_assert` will trigger even if the code does not run. On the other hand, if I use a template it will delay the evaluation.

Reply:
I suppose that this should be solved differently - could you tell me where you use it?

Reply:
It was supposed to be used when I push the next pull request.

Reply:
It remained there when I was removing the static functions.

Reply:
> The problem with `std::false_type` is it does not delay the `static_assert`.

Why do you want to delay the `static assert`? An idea on how do you plan to use it will be enough.

> If I use `std::false_type` `static_assert` will trigger even if the code does not run.

Yes, Static assert is compile-time, why not use a runtime `assert` then? Again things will be more clear if you could share the use case.

Reply:
Its use was removed right now as I wrote earlier it was remains of struct or functions which forgot to remove once @bassoy reviews the new change I will remove it. For cases, I can give you some.
- let's say you have templated functions or structs as of C++17 we don't have concepts which will make the error propagate through the whole instantiation which produces the whole unnecessary error and hard to find the error. So this helps to stop the error because when it hit `static_assert` it will always false stop there and also helps with the better error system.
```cpp
/// Make the user clear about the use of this struct and improves the error message.
template<typename T>
struct Integer{
static_assert(always_false_v<T>,"type does not satisfy the condition");
}

template<>
struct Integer<int>{};

```
- If we have a very large `constexpr if` it will help to satisfy the DRY principle. It also acts as barrier which makes sure you never reach there.
```cpp
if consexpr(....){}
else if consexpr(....){}
....
else{
static_assert(always_false_v<T>,"");
}
```

Reply:
The trait `always_false` is too general and the naming does not provide enough information what is wrong about the type. The information should not be provided with a string.

Reply:
I have removed it from lib.